pennylane/devices/default_qubit.py

```diff
@@ -436,6 +446,37 @@ def _apply_phase(self, state, axes, parameters, inverse=False):
         phase = self._conj(parameters) if inverse else parameters
         return self._stack([state[sl_0], phase * state[sl_1]], axis=axes[0])
 
+    def expval(self, observable, shot_range=None, bin_size=None):
+        """Returns the expectation value of a Hamiltonian observable. When the observable is a
+         `SparseHamiltonian` object, the expectation value is computed directly, which leads to
+          faster execution.
+
+        Args:
+            observable (~.Observable): a PennyLane observable
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+
+        Returns:
+            float[real]/callable: returns the expectation value of the observable if it is a sparse
+            matrix in scipy coordinate list (COO) format or returns the :func:`~.expval` function
+            from the parent class
+        """
+        if observable.name == "SparseHamiltonian" and self.shots is None:
+
+            ev = coo_matrix.dot(
+                coo_matrix(self.state),
+                coo_matrix.dot(
+                    observable.matrix, coo_matrix(self.state.reshape(len(self.state), 1))
+                ),
+            )
+
+            return np.real(ev.toarray()[0])
```

Comment:
I believe this requires having access to the sparse hamiltonian eigenbasis which requires diagonalizing the matrix. If that is correct, then yes, we can add such functionality in a future PR.

Reply:
To sample from the full Hamiltonian, you would need to diagonalize it. This should be very time-consuming and unlikely to generate impressive speedups. So it's possible, but probably not worth it

Reply:
Thanks @ixfoduap, good to know. Out of curiosity, what happens at the moment if you attempt to do:

- `return qml.var(qml.SparseHamiltonian(...))` with `shots=None`/`shots=100`
- `return qml.sample(qml.SparseHamiltonian(...))`. 

Do you get a useful error message?

Reply:
@josh146 This is what we currently get:
```
~/Soran/Xanadu/Edit/pennylane/pennylane/operation.py in _eigvals(cls, *params)
    945             array: eigenvalue representation
    946         """
--> 947         raise NotImplementedError
    948 
    949     @property

NotImplementedError: 
```

I can add conditions to the `sample` and `var` functions in `_qubit_device.py` or I can create sample and var branches in `default_qubit.py` that raise errors when the observable is a sparse Hamiltonian, otherwise they return the superclass functions (similar to what we did for expval). Any thoughts?

Reply:
Thanks @josh146. I agree that creating two branches just for the sake of raising errors is not very clean. I am in favour of explaining things in the docstring for now until we collect all sparse functions in one place where we can raise appropriate error messages.